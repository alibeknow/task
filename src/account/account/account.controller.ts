import { Controller, Inject } from '@nestjs/common';
import { Ctx, EventPattern, Payload, RmqContext } from '@nestjs/microservices';
import { ServiceEvents } from '@shared/microservices';
import { AccountService } from '@shared/account';
import { AccountDto } from './account.dtos';

@Controller()
export class AccountController {
  constructor(private readonly accountService: AccountService) {}

  @EventPattern(ServiceEvents.GET_ACCOUNT)
  async eventHandler(
    @Ctx() context: RmqContext,
    @Payload() eventData: AccountDto,
  ) {
    await this.accountService.findOne(eventData.accountId);
    const channel = context.getChannelRef();
    channel.ack(context.getMessage());
  }
}
