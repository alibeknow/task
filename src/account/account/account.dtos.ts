import { IsString } from 'class-validator';

export class AccountDto {
  @IsString()
  public readonly accountId!: string;
}
