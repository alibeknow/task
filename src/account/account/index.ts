export * from './account.controller';
export * from './account.dtos';
export * from './account.module';
export * from './account.service';
