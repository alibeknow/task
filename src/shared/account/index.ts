export * from './account.entity';
export * from './account.service';
