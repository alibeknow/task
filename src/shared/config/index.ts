export * from './api-config.service';
export * from './config.module';
