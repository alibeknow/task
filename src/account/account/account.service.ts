import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { AccountEntity } from '../../shared/account/account.entity';
import { Repository } from 'typeorm';
import { AccountDto } from './account.dtos';

@Injectable()
export class AccountService {
  constructor(
    @InjectRepository(AccountEntity)
    private accountRepository: Repository<AccountEntity>,
  ) {}

  getAccount(eventData: AccountDto): Promise<AccountEntity | null> {
    return this.accountRepository.findOne({
      where: {
        id: eventData.accountId,
      },
    });
  }
}
