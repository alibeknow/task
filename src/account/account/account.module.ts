import { Module, forwardRef } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { ConfigModule as CustomConfigModule } from '@shared/config';
import { AccountController } from './account.controller';
import { AccountService } from '@shared/account';
import { TypeOrmModule } from '@nestjs/typeorm';
import { AccountEntity } from '@shared/account';
import { TransactionModule } from '@transaction/transaction';
import {
  MESSAGE_BUS_PROVIDER,
  clientFactory,
  RABBIT_TRANSACTION_SEND_QUEUE,
} from '@shared/microservices';

@Module({
  imports: [
    TypeOrmModule.forFeature([AccountEntity]),
    CustomConfigModule,
    forwardRef(() => TransactionModule),
  ],
  controllers: [AccountController],
  providers: [
    {
      provide: RABBIT_TRANSACTION_SEND_QUEUE,
      useFactory: (configService: ConfigService) =>
        configService.get<string>(RABBIT_TRANSACTION_SEND_QUEUE),
      inject: [ConfigService],
    },
    {
      provide: MESSAGE_BUS_PROVIDER,
      useFactory: clientFactory,
      inject: [ConfigService, RABBIT_TRANSACTION_SEND_QUEUE],
    },
    AccountService,
  ],
  exports: [AccountService],
})
export class MicroserviceAccountModule {}
