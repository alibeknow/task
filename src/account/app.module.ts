import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { AccountModule } from './account';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
    }),
    AccountModule,
  ],
})
export class AppModule {}
