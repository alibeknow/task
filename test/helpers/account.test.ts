import { DatabaseModule } from '@shared/db';
import { Test } from '@nestjs/testing';
import { HttpServer, INestApplication } from '@nestjs/common';
import request from 'supertest';
import { getNestApp } from './nest.app';
import { Connection, getConnection } from 'typeorm';
import { AccountEntity } from '../../src/shared/account/account.entity';
import { AppGatewayModule } from '../../src/gateway/app.module';
import { ConfigModule } from '@nestjs/config';

async function insertDataAccount(connection: Connection) {
  const accounts = [
    {
      id: '1c7b3d22-bfe7-4fcb-8c00-f77039da065b',
      createdAt: '2022-07-11T05:22:49.029Z',
      updatedAt: '2022-07-11T05:22:49.029Z',
      firstName: 'firstAccount',
      lastName: 'firstAccount',
      balance: 40000,
    },
    {
      id: '20863556-645c-4db4-a2be-cabcbe324f99',
      createdAt: '2022-07-11T05:22:49.029Z',
      updatedAt: '2022-07-11T05:22:49.029Z',
      firstName: 'secondaccount',
      lastName: 'secondaccount',
      balance: 10000,
    },
    {
      id: '3592e667-afee-4b2b-9e73-2e228ba047e8',
      createdAt: '2022-07-11T05:22:49.029Z',
      updatedAt: '2022-07-11T05:22:49.029Z',
      firstName: 'thirdaccount',
      lastName: 'thirdaccount',
      balance: 0,
    },
  ];
  connection
    .createQueryBuilder()
    .insert()
    .into(AccountEntity)
    .values(accounts)
    .execute();
}

describe('[Metadata endpoints]', () => {
  let app: INestApplication;
  let httpServer: HttpServer;
  let connection: Connection;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [AppGatewayModule,ConfigModule,DatabaseModule],
    }).compile();

    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();
    connection = await getConnection(); 
  });

  describe('[GET account]', () => {
    it('should return 404 error if requested item doesn`t exist in our DB', async () => {
      const actualResult = await request(httpServer)
        .get('/blahblah')
        .set('Accept', 'application/json');
      console.log(actualResult)
      expect(actualResult.statusCode).toBe(404);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return 400 error if validation doesn`t pass', async () => {
      const actualResult = await request(httpServer)
      .get('/account/heasd')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct result based on request params', async () => {
      await insertDataAccount(connection);

      const actualResult1 = await request(httpServer)
        .get('/account/1c7b3d22-bfe7-4fcb-8c00-f77039da065b')
        .set('Accept', 'application/json'); 
        console.log(actualResult1)

      const actualResult2 = await request(httpServer)
        .get('/account/20863556-645c-4db4-a2be-cabcbe324f99')
        .set('Accept', 'application/json');
      
      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();
    });
  });

  afterEach(async () => {
      await connection
        .createQueryBuilder()
        .delete()
        .from(AccountEntity)
        .execute();
  });

  afterAll(async () => {
    await connection.close();
    await app.close();
  });
});
