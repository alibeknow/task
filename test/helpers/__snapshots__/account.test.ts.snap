// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Metadata endpoints] [GET account] should return 400 error if validation doesn\`t pass 1`] = `
Object {
  "error": "Bad Request",
  "message": "Validation failed (uuid  is expected)",
  "statusCode": 400,
}
`;

exports[`[Metadata endpoints] [GET account] should return 404 error if requested item doesn\`t exist in our DB 1`] = `
Object {
  "error": "Not Found",
  "message": "Cannot GET /blahblah",
  "statusCode": 404,
}
`;

exports[`[Metadata endpoints] [GET account] should return correct result based on request params 1`] = `
Object {
  "result": Object {
    "balance": 40000,
    "createdAt": "2022-07-11T05:22:49.029Z",
    "firstName": "firstAccount",
    "id": "1c7b3d22-bfe7-4fcb-8c00-f77039da065b",
    "lastName": "firstAccount",
    "updatedAt": "2022-07-11T05:22:49.029Z",
  },
  "statusCode": 200,
}
`;

exports[`[Metadata endpoints] [GET account] should return correct result based on request params 2`] = `
Object {
  "result": Object {
    "balance": 10000,
    "createdAt": "2022-07-11T05:22:49.029Z",
    "firstName": "secondaccount",
    "id": "20863556-645c-4db4-a2be-cabcbe324f99",
    "lastName": "secondaccount",
    "updatedAt": "2022-07-11T05:22:49.029Z",
  },
  "statusCode": 200,
}
`;
